### ISSUE TITLE
Authentication failed with RSA key when testing Example #2 of sshtunnel

### ISSUE THREAD

**Issue by ejm4567** [paramiko/paramiko#2109]
I am getting an Authentication error when I try running a test script that implements Example #2 of the sshtunnel documentation.

The script is

```
with sshtunnel.open_tunnel(
        ssh_address_or_host=(GATEWAY_SERVER_IP, GATEWAY_PORT),
        ssh_username=SSH_UNAME,
        ssh_pkey=SSH_PKEY_FILENAME,
        ssh_private_key_password=SSH_PW,
        local_bind_address=('0.0.0.0', 5000),
        remote_bind_address=(REMOTE_SERVER_IP, REMOTE_PORT),
        debug_level=DEBUG_LEVEL
        ) as tunnel:
    tunnel.start()
    client = paramiko.SSHClient()
    client.load_system_host_keys()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    client.connect('127.0.0.1', 5000)
    client.exec_command('echo I am the client...')
    client.close()
```

where the variables in upper case are all passed in from the command line and are either IP addresses or port nos. as integers.

Python 3.7.11 / sshtunnel 0.4.0 / paramiko 2.11.0 / OpenSSL 1.1.1k  FIPS 25 Mar 2021 / CentOS Linux release 8.5.2111

The error I see with DEBUG_LEVEL = 1 is:

```
[emansky@sdo8 PythonScripts]$ ./test_tunnelClient.py -g=A.B.C.D -gp=22 -r=A.B.C.E -rp=22 -name emansky -f /home/emansky/.ssh/id_rsa -pw PASSPHRASE -d 1
INF | MainThrea/1060@sshtunnel | 3 keys loaded from agent
ERR | MainThrea/1315@sshtunnel | Password is required for key /home/emansky/.ssh/id_rsa
INF | MainThrea/1117@sshtunnel | 3 key(s) loaded
DEB | MainThrea/1310@sshtunnel | Private key file (/home/emansky/.ssh/id_rsa, <class 'paramiko.rsakey.RSAKey'>) successfully loaded
INF | MainThrea/0981@sshtunnel | Connecting to gateway: A.B.C.D:22 as user 'emansky'
DEB | MainThrea/0984@sshtunnel | Concurrent connections allowed: True
DEB | MainThrea/1401@sshtunnel | Trying to log in with key: b'84b1e63b918139d9479e96c032499da6'
DEB | MainThrea/1205@sshtunnel | Transport socket info: (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 0), timeout=0.1
2022-08-26 11:31:08,026| DEB | MainThrea/1410@sshtunnel | Authentication error
2022-08-26 11:31:08,026| WAR | MainThrea/1450@sshtunnel | Tunnels are not started. Please .start() first!
2022-08-26 11:31:08,026| INF | MainThrea/1474@sshtunnel | Closing ssh transport
2022-08-26 11:31:08,026| DEB | MainThrea/1477@sshtunnel | Transport is closed
2022-08-26 11:31:08,026| DEB | MainThrea/1401@sshtunnel | Trying to log in with key: b'f772323c3db69fa96eb8dcd31d0174eb'
2022-08-26 11:31:08,097| DEB | MainThrea/1205@sshtunnel | Transport socket info: (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 0), timeout=0.1
DEB | MainThrea/1410@sshtunnel | Authentication error
WAR | MainThrea/1450@sshtunnel | Tunnels are not started. Please .start() first!
INF | MainThrea/1474@sshtunnel | Closing ssh transport
DEB | MainThrea/1477@sshtunnel | Transport is closed
DEB | MainThrea/1401@sshtunnel | Trying to log in with key: b'111ef38efbbe57c5471fe0ed25832e65'
DEB | MainThrea/1205@sshtunnel | Transport socket info: (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 0), timeout=0.1
ERR |  Thread-5/1874@transport | Exception (client): key cannot be used for signing
ERR |  Thread-5/1872@transport | Traceback (most recent call last):
ERR |  Thread-5/1872@transport |   File "/usr/local/lib/python3.7/site-packages/paramiko/transport.py", line 2164, in run
ERR |  Thread-5/1872@transport |     handler(self.auth_handler, m)
ERR |  Thread-5/1872@transport |   File "/usr/local/lib/python3.7/site-packages/paramiko/auth_handler.py", line 395, in _parse_service_accept
ERR |  Thread-5/1872@transport |     sig = self.private_key.sign_ssh_data(blob, algorithm)
ERR |  Thread-5/1872@transport |   File "/usr/local/lib/python3.7/site-packages/paramiko/agent.py", line 436, in sign_ssh_data
ERR |  Thread-5/1872@transport |     raise SSHException("key cannot be used for signing")
ERR |  Thread-5/1872@transport | paramiko.ssh_exception.SSHException: key cannot be used for signing
ERR |  Thread-5/1872@transport |
ERR | MainThrea/1223@sshtunnel | Could not connect to gateway A.B.C.D:22 : key cannot be used for signing
Traceback (most recent call last):
  File "./test_tunnelClient.py", line 51, in <module>
    debug_level=DEBUG_LEVEL
  File "/usr/local/lib/python3.7/site-packages/sshtunnel.py", line 1608, in __enter__
    self.start()
  File "/usr/local/lib/python3.7/site-packages/sshtunnel.py", line 1332, in start
    reason='Could not establish session to SSH gateway')
  File "/usr/local/lib/python3.7/site-packages/sshtunnel.py", line 1174, in _raise
    raise exception(reason)
sshtunnel.BaseSSHTunnelForwarderError: Could not establish session to SSH gateway
```

I tried allow_agent = False too, with the same result.
I also converted the RSA key to PEM encoding, again with the same result, no connection.

Any tips or ideas would be greatly appreciated!

Thanks,

--Ed

**Comment by ejm4567** [comment]
I've resolved the issue and was able to get Example #2 to work. Turns out the issue was that the local ssh agent did not have loaded the RSA key I was wanting to use. So my current RSA key was never used, and the passphrase passed in did not match the old keys (I had changed my passphrase). Hence the error message from paramiko was indeed valid.

Deleting the existing keys in the agent, then adding my current key only to the agent then resulted in the key being valid for signing and a tunnel was established.

For anyone else having similar issues, here's what the output looks like with DEBUG_LEVEL = 1 (trace):

```
./test_tunnelClient2.py -g A.B.C.D -gp 22 -r A.B.C.E -rp 22 -d 1 -f ~/.ssh/id_rsa -pw PASSWORD -ppw PASSPHRASE

2022-09-19 12:08:49,263| INF | MainThrea/1060@sshtunnel | 1 keys loaded from agent\
2022-09-19 12:08:49,264| ERR | MainThrea/1314@sshtunnel | Password is required for key /home/emansky/.ssh/id_rsa\
2022-09-19 12:08:49,264| DEB | MainThrea/1309@sshtunnel | Private key file (/home/emansky/.ssh/id_ed25519, <class 'paramiko.ed25519key.Ed25519Key'>) successfully loaded\
2022-09-19 12:08:49,264| INF | MainThrea/1117@sshtunnel | 2 key(s) loaded\
2022-09-19 12:08:49,484| DEB | MainThrea/1309@sshtunnel | Private key file (/home/emansky/.ssh/id_rsa, <class 'paramiko.rsakey.RSAKey'>) successfully loaded\
2022-09-19 12:08:49,485| INF | MainThrea/0978@sshtunnel | Connecting to gateway: A.B.C.D:22 as user 'emansky'\
2022-09-19 12:08:49,485| DEB | MainThrea/0983@sshtunnel | Concurrent connections allowed: True\
2022-09-19 12:08:49,485| DEB | MainThrea/1400@sshtunnel | Trying to log in with key: b'eeb2d32b9fd11f6a7515b27b2fe60c1a'\
2022-09-19 12:08:49,559| DEB | MainThrea/1204@sshtunnel | Transport socket info: (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 0), timeout=0.1\
KT =  ssh-rsa
2022-09-19 12:08:50,194| DEB | MainThrea/1410@sshtunnel | Authentication error\
2022-09-19 12:08:50,194| WAR | MainThrea/1450@sshtunnel | Tunnels are not started. Please .start() first!\
2022-09-19 12:08:50,194| INF | MainThrea/1474@sshtunnel | Closing ssh transport\
2022-09-19 12:08:50,194| DEB | MainThrea/1477@sshtunnel | Transport is closed\
2022-09-19 12:08:50,194| DEB | MainThrea/1400@sshtunnel | Trying to log in with key: b'eeb2d32b9fd11f6a7515b27b2fe60c1a'\
2022-09-19 12:08:50,263| DEB | MainThrea/1204@sshtunnel | Transport socket info: (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 0), timeout=0.1
KT =  ssh-rsa
rsa-sha2-512
14
2022-09-19 12:08:51,425| DEB | MainThrea/1410@sshtunnel | Authentication error\
2022-09-19 12:08:51,425| WAR | MainThrea/1450@sshtunnel | Tunnels are not started. Please .start() first!\
2022-09-19 12:08:51,425| INF | MainThrea/1474@sshtunnel | Closing ssh transport\
2022-09-19 12:08:51,425| DEB | MainThrea/1477@sshtunnel | Transport is closed\
2022-09-19 12:08:51,425| DEB | MainThrea/1400@sshtunnel | Trying to log in with key: b'c0d767fb0ee5f920e16163efa9bb0405'\
2022-09-19 12:08:51,495| DEB | MainThrea/1204@sshtunnel | Transport socket info: (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 0), timeout=0.1\
KT =  ssh-ed25519
2022-09-19 12:08:52,120| DEB | MainThrea/1410@sshtunnel | Authentication error\
2022-09-19 12:08:52,120| WAR | MainThrea/1450@sshtunnel | Tunnels are not started. Please .start() first!\
2022-09-19 12:08:52,120| INF | MainThrea/1474@sshtunnel | Closing ssh transport\
2022-09-19 12:08:52,120| DEB | MainThrea/1477@sshtunnel | Transport is closed\
2022-09-19 12:08:52,120| DEB | MainThrea/1414@sshtunnel | Trying to log in with password: *******\
2022-09-19 12:08:52,211| DEB | MainThrea/1204@sshtunnel | Transport socket info: (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 0), timeout=0.1\
2022-09-19 12:08:52,856| INF |  Srv-5000/1433@sshtunnel | 
    **Opening tunnel: 0.0.0.0:5000 <> A.B.C.E:22**

2022-09-19 12:08:52,856| INF | MainThrea/0793@sshtunnel | Checking tunnel to: ('A.B.C.E, 22)\
2022-09-19 12:08:52,928| TRA | Thread-9 /0360@sshtunnel | #1 <-- ('127.0.0.1', 60444) connected\
2022-09-19 12:08:53,037| TRA | Thread-9 /0333@sshtunnel | <<< IN #1 <-- ('127.0.0.1', 60444) recv: b'5353482d322e302d4f70656e5353485f382e300d0a' <<<\
2022-09-19 12:09:03,856| DEB | MainThrea/0817@sshtunnel | 
    **Tunnel to ('A.B.C.E', 22) is UP**
2022-09-19 12:09:03,857| TRA | Thread-9 /0368@sshtunnel | #1 <-- ('127.0.0.1', 60444) sending RST\
2022-09-19 12:09:03,857| TRA | Thread-9 /0375@sshtunnel | #1 <-- ('127.0.0.1', 60444) connection closed.\
2022-09-19 12:09:03,998| TRA | Thread-11/0360@sshtunnel | #2 <-- ('127.0.0.1', 60446) connected\
2022-09-19 12:09:03,998| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'5353482d322e302d706172616d696b6f5f322e31312e300d0a' >>>\
2022-09-19 12:09:04,067| TRA | Thread-11/0333@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv: b'5353482d322e302d4f70656e5353485f382e300d0a' <<<\
2022-09-19 12:09:04,067| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'0000048c0514282c87942c41fbe8dfa0fb26efdba8f40000011b637572766532353531392d736861323536406c69627373682e6f72672c656364682d736861322d6e697374703235362c656364682d736861322d6e697374703338342c656364682d736861322d6e697374703532312c6469666669652d68656c6c6d616e2d67726f757031362d7368613531322c6469666669652d68656c6c6d616e2d67726f75702d65786368616e67652d7368613235362c6469666669652d68656c6c6d616e2d67726f757031342d7368613235362c6469666669652d68656c6c6d616e2d67726f75702d65786368616e67652d736861312c6469666669652d68656c6c6d616e2d67726f757031342d736861312c6469666669652d68656c6c6d616e2d67726f7570312d736861312c6578742d696e666f2d630000018b7373682d656432353531392c65636473612d736861322d6e697374703235362c65636473612d736861322d6e697374703338342c65636473612d736861322d6e697374703532312c7273612d736861322d3531322c7273612d736861322d3235362c7373682d7273612c7373682d6473732c7373682d656432353531392d636572742d763031406f70656e7373682e636f6d2c65636473612d736861322d6e697374703235362d636572742d763031406f70656e7373682e636f6d2c65636473612d736861322d6e697374703338342d636572742d763031406f70656e7373682e636f6d2c65636473612d736861322d6e697374703532312d636572742d763031406f70656e7373682e636f6d2c7273612d736861322d3531322d636572742d763031406f70656e7373682e636f6d2c7273612d736861322d3235362d636572742d763031406f70656e7373682e636f6d2c7373682d7273612d636572742d763031406f70656e7373682e636f6d2c7373682d6473732d636572742d763031406f70656e7373682e636f6d0000004a6165733132382d6374722c6165733139322d6374722c6165733235362d6374722c6165733132382d6362632c6165733139322d6362632c6165733235362d6362632c336465732d6362630000004a6165733132382d6374722c6165733139322d6374722c6165733235362d6374722c6165733132382d6362632c6165733139322d6362632c6165733235362d6362632c336465732d63626300000083686d61632d736861322d3235362c686d61632d736861322d3531322c686d61632d736861322d3235362d65746d406f70656e7373682e636f6d2c686d61632d736861322d3531322d65746d406f70656e7373682e636f6d2c686d61632d736861312c686d61632d6d64352c686d61632d736861312d39362c686d61632d6d64352d393600000083686d61632d736861322d3235362c686d61632d7368' >>>\
2022-09-19 12:09:04,067| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'61322d3531322c686d61632d736861322d3235362d65746d406f70656e7373682e636f6d2c686d61632d736861322d3531322d65746d406f70656e7373682e636f6d2c686d61632d736861312c686d61632d6d64352c686d61632d736861312d39362c686d61632d6d64352d3936000000046e6f6e65000000046e6f6e65000000000000000000000000000000000000' >>>\
2022-09-19 12:09:04,175| TRA | Thread-11/0333@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv: b'00000414051450269b71dcb0f30641d697ba738f2d2300000125637572766532353531392d7368613235362c637572766532353531392d736861323536406c69627373682e6f72672c656364682d736861322d6e697374703235362c656364682d736861322d6e697374703338342c656364682d736861322d6e697374703532312c6469666669652d68656c6c6d616e2d67726f75702d65786368616e67652d7368613235362c6469666669652d68656c6c6d616e2d67726f757031342d7368613235362c6469666669652d68656c6c6d616e2d67726f757031362d7368613531322c6469666669652d68656c6c6d616e2d67726f757031382d7368613531322c6469666669652d68656c6c6d616e2d67726f75702d65786368616e67652d736861312c6469666669652d68656c6c6d616e2d67726f757031342d73686131000000417273612d736861322d3531322c7273612d736861322d3235362c7373682d7273612c65636473612d736861322d6e697374703235362c7373682d65643235353139000000776165733235362d67636d406f70656e7373682e636f6d2c63686163686132302d706f6c7931333035406f70656e7373682e636f6d2c6165733235362d6374722c6165733235362d6362632c6165733132382d67636d406f70656e7373682e636f6d2c6165733132382d6374722c6165733132382d636263000000776165733235362d67636d406f70656e7373682e636f6d2c63686163686132302d706f6c7931333035406f70656e7373682e636f6d2c6165733235362d6374722c6165733235362d6362632c6165733132382d67636d406f70656e7373682e636f6d2c6165733132382d6374722c6165733132382d636263000000a9686d61632d736861322d3235362d65746d406f70656e7373682e636f6d2c686d61632d736861312d65746d406f70656e7373682e636f6d2c756d61632d3132382d65746d406f70656e7373682e636f6d2c686d61632d736861322d3531322d65746d406f70656e7373682e636f6d2c686d61632d736861322d3235362c686d61632d736861312c756d61632d313238406f70656e7373682e636f6d2c686d61632d736861322d353132000000a9686d61632d736861322d3235362d65746d406f70656e7373682e636f6d2c686d61632d736861312d65746d406f70656e7373682e636f6d2c756d61632d3132382d65746d406f70656e7373682e636f6d2c686d61632d736861322d3531322d65746d406f70656e7373682e636f6d2c686d61632d736861322d3235362c686d61632d736861312c756d61632d313238406f70656e7373682e636f6d2c686d61632d736861322d353132000000156e6f6e652c7a6c6962406f70656e7373682e636f6d000000156e6f6e652c7a6c6962406f70656e73' <<<\
2022-09-19 12:09:04,175| TRA | Thread-11/0333@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv: b'73682e636f6d000000000000000000000000000000000000' <<<\
2022-09-19 12:09:04,218| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'0000002c061e000000207fb0a68613f0acf2e98069c3f1d70a10dfd65f67f757b533f2fb817a4b1ee778000000000000' >>>\
2022-09-19 12:09:04,297| TRA | Thread-11/0333@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv: b'000000bc081f000000330000000b7373682d65643235353139000000208faffb1bd8e795beb0a1e6be4dfe20ded3f7ab86d7ed05d93d061200d95a75e500000020252a5076adcaaac8ca1dc08c180ee095390fed06e8c4dce5a155cc5fdec8f569000000530000000b7373682d6564323535313900000040320c4b81ef546d0902ee20cf0a35933d7266100226c3fa5d30eccd8d72cd895f562b0db342378e0cbbc4dbeb227d8f27b1cd530a5c074938df1525e8a33a760900000000000000000000000c0a150000000000000000000074b1d856dbcf4a78311a74795dfd5a1ab3f25cc6e05b55c2d1c86f27d2f72006ce61289654f0164183d8bd016c42e4947bd448d231f14834a6cc4b0886c1f0f0c8574cdfd609293b8812cd04fb5ab359d7b4afeff1c55c5a86928d6ea44f9a19f103d89db061fc4f9509355f4b76cf51b7fe1a1e4bc1ba2a42366b98b88ba4ae3975fd2b7a534c8615aa5540db89baf89a387b8bd221043340ff2926e6e092289b6c70b98c2e4051644010c092f14ca238e5d2b0a7aa63ee807809444be6cd2e' <<<\
2022-09-19 12:09:04,298| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'0000000c0a1500000000000000000000' >>>\
2022-09-19 12:09:06,929| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'5cf6a35afe9b80e29aba1f834f4059dc0eabe8a27af87ab0da91ae14c8991c23a3cec94884147e397167368ea810d4b855c8aa55dd1584007fc48b794450e551' >>>\
2022-09-19 12:09:07,008| TRA | Thread-11/0333@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv: b'8257c8e82e3fef6545beb037511ba84a3acfad789d8694335248973c19c3d694ca9c9933792e975f7bffc382da58f616cd5e99d3e3deb32cb9b608c487fc260c' <<<\
KT =  ssh-rsa
rsa-sha2-512
14
2022-09-19 12:09:07,410| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'4fe7d5165c0bb819d7b6c535d699655f5b9c10bfde38ba42452e50edcbd233035e1c2ddfc14027e8b622310889901434421a7978df148ff0d9a55a2c6b459764dfb6bbb06f31e0b0473d6edc971d48fcd95c2591cb11245cbd7034a15ba9b89a73c3eb579d231131bc72cf3e2c87aa9814303584a6569693fb912712b3a2aebf303013e63c263a590d1dfb520e3efc3bfc90468d4cc128bd9f0a1ca2283ababa37df9369f0ba886dff50ebe4d7d7f1948cc0ad7470757459bbfaf92a1a5b9032302397a610cf9daa2d3779bb10505adcf895e83d2c9ed4580290289f55df073955afecaa2b69550d986b044ea69eb27ea49d9c117c55736d17234889ad415f71d7059fd7e2350c4f5dc8245a6b5e21ca9fe199b5bdb9983dfc6fbe398ef3ac515ca62d32a5755c8634dc9b65cb0f4e44203497aabff61727fa6764088995bbdc8bc96abbdc6d64931553cc64964349c9a729860289ae57ff13e0d02a70a6e1055a21f9cd49682d39953812f6d9f4d8769515b1c74254d506e3ee7b3da26abc8a3d4d623d7804b1476509bd71032c7ae342d2de74e7385bbbd85c3c521247c8d41b0e7d563c304e3974ed142105b09e0d01aa4464625ff520f0d228f0cba3b90ae35bdbb9b905b96d133e53ea5ba083f7119986b62d0488f878113bd34e0e624077bebef1c0571b28adc465036b7d9bd7133dcb17da3b78cdc2b21935143aed7e9642aef31eda4e48690f758f40b3314d88df47dd8fd2fd1ed91db103ce3d6c8d1819b2004a35334c9ab027e9eefef282cdf7b7c5461de178a8dd2621365be5040161a6324a4359e390e7a2a17ad0f686b626c1ef4e9dd38d373d9ad41bd30fc2185a8afb056dc1220f82a64ca46f1c3902a080bd922d22ae01a26cf02b2bae64e852aefb4377afc6733a1546322bdcf5698a570a3763ae07b8f2c0f7ecbb40fead8044216417978398a19bf9448001fae93853a60c00dd231ec2519183160adcf6cca5c85004fbad12c48da8d53384cab457cbfcc3dc4a129e864ec7d0340a3c120aab98ab3dbf0c1db44885873650b42003b567db612f209c631e4408a5ae246cc56b9df6416109a33ebeafb5e71562a70d8d5a6a022180e2ffba468484425292ae6f9d40510ae1a3dfc8405004d904f5e09b69d595cf26582f4c949708ac9f9dee1451d3e4486f44c29c701d8446d5b7fe29979e54e56202c8fb233daf1baaab4a08c29ac53aef1f7c699af4212f660199857e8c963e818eac62a5fb83f4ac1d1f48872c9a3d573da5c66ed4cb6cf1f5011bec8820da7b97e551ddf473898b' >>>\
2022-09-19 12:09:07,486| TRA | Thread-11/0333@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv: b'39697606f42bea30c9a581cf43722ed78fa9685f02ee00d1d82e44dc54404021c9b97ad9548ce0f6ed13235db21f7ff4' <<<\
2022-09-19 12:09:07,486| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'be548457fcdb669a21ef7251ba3522eaab3e486490d8aee3566e0da9d03793bc68fca2af889ac20e588fcf08ca5e8626d0888d9a94cb9e163f661366feb4defcbfc697e89fb9e09828539bc2f3319c1e' >>>\
2022-09-19 12:09:07,575| TRA | Thread-11/0333@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv: b'cbe02b094e5a07cbd3e079b7ba8dc2c0bca921511f943d6b1d18d66ff02392567fc3235e439110f2acb5c5eb3231cb815e2b1b30424b81469626e4a1ea48d1197017f131995405dc54b62cc4100fe5f08cdb4e4c03fef5e26214d254bf3fbce4cb16ff102861d0f1260d369822482a9e7c15263d39baeb675ded3cef6a7ee413bbfdc80a2380733b2e67fb7f36ade1efefd7367e3471f978b7a2561f215effb79bcdbb6d12a22c39024931e0fb8729dda7cad50b3db5256b9996077fd9c6f8157521e2f3aafeba42730e1b10ef84f25baba1a2290bb1fe89ef94c5c67430a64d15fc8b669e3a5b66610d1150ab5ff0cc18526aa64d86c0fd6b2d97bf1328ba1aa72acd1c07c98636da773a1ad6d3fbe625056af92899fc302ef62f63b02c7a62ada33bed572519a8d9ec93e3b12a4134f597dd3ea10758900a8ee27804b2d6b86182a74a964af756354af95dcc8cb6b1cf9ae5314165c47b106da8062a2aeba34952dd9a9a28daacbd71812ac827e062a7f255b5ec50f164709cda351386499b64f813707cdcad8b8e6094284162bacdb539b29db05c5e401d346c8fcda57e9a39bbe31a57cd113980c13541e351a2043cd9cdf6f4a71e044b9a70a8efbfb50acc3a391c812ed05022136bd91fc8a14ff72f607b1fbc19f061f2841537f20d96a42f2e143169b8110c02fab3be41abbc2ec285615db764611d2ded7d179b0b60c95faba871dcdbde7ffe4c3dfd73d8be43a67ba4caa332cf412a17e96711a41425892998d890a8b424fed3f3a3790362039c6b98c995f7a85b6e23dbc57ba09ab424dc1262668c17f8c3d7d39efc43bbfb13b909318777e1d12180c527abacc932a88307dda4e94f99e52532ae58ecd0bfe439f5ce36349c39b716b87b1be0f6cc6293317536114c81045f26ed15ddcbd5f4ed8828cbf0c69237074475be82df77148e11bc2c4b6d09f505a0f319221c0088d22aaeefb52305f0c34bbc850db86fe83073416b681ee40cce765c6838b956bbd1533a6b9b84858120c59886e6630ab8330b345cc0797467c61e98ea956e5f70f4d6b6c373301803f66f076094e1aabc638830af9c468f3342dfc4b80decc3032bbbf6e4d133fb4f7ad7482f346fd70bc4670b0c823ebfb8e317e27b89eb4a12151f78fffcad9c0fe45fbf984d7f' <<<\
2022-09-19 12:09:07,684| TRA | Thread-11/0333@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv: b'e5c8bd589744478a9c668f61964931a277ff4fa6ba9ed1a1813e45f5cd38c3b21befd5bce4853b4b9ccf477b967de1b468506e192f9e78e31785262ae7aa1441' <<<\
2022-09-19 12:09:07,685| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'119861d578e6862adbbdbdfb81b822484089492ba0b9d6df92fb944b3d8096a2906df58c3162f29c5b10c166e091b3dce419b50bab314a91880b607b55c652ab85b7033331543deb3ec8f90ef97a207f' >>>\
2022-09-19 12:09:07,755| TRA | Thread-11/0333@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv: b'65c25a2ff674f251951df0a03b3e75336006230113f1a883084bf963adc251d0c34bc08926a8e039a452d0b8e61ca4af0ceede154d253ae32777202f7008db908c65d04f9518e9a3b70c81310814a0365f4b6cf04389c74d637440d27eb4e14fbfbd92932ad34872b3798535660d2112' <<<\
2022-09-19 12:09:07,864| TRA | Thread-11/0333@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv: b'37b0ca6e7ff3d65ab10c9617d046201338fe829c53ef666b2e5ffcd8b2eaa015d15f2dd349207fc828f4d80a7191f5960d39b6ecee5502b080ec3185b0f74b374d78bd33c00a0b8983177226d1e4033f6f84fc87683295fa55bb2cda9eb06c201d5d671a6116b543b2840e6e7cfa1decb0045e82de31af4bbb39ef3a7aadfbc7ab533908b6e84b5299fe8ec8da0d95890f259c553cc4ba68367f1b062fbaca05849d36b9980c629943844a17cbfaeb21aa46b268b1654c63acc2d3b5d9a20f45c90d07f8fadfde792f464e0d7130f39fdcfa984ba910e932ed6ac8d71fa7edb2c6c8bd13837c8c290d0184ec6088c6ca27fc2bfbf8765461831c4511397feb99a697632086c239d23d0476642d680e45636d20453f79d0dbd6a108ff9b5c14a2c0f563c619b2f7dc54faf1e3af86dae25ce01f3f624b05567c314fbbedfa29307c4abd04ab0abe24cda015549c0718a5aff1ff732f3b862fee066ba14d2329e337c8bbcf4551ef38bb433c5880a6eb3b' <<<\
2022-09-19 12:09:07,865| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'f2140552dfd1b9f11e9bb58f03652b435c9fb34b5a0fee4d2cf212056a6001fa9013a4fb3274c4378cbad03953f55d7d' >>>\
2022-09-19 12:09:07,865| TRA | Thread-11/0316@sshtunnel | >>> OUT #2 <-- ('127.0.0.1', 60446) send to ('A.B.C.E', 22): b'708c62bf07bd0b6a5498138779a8fbd7dcca15dc865b55533780e35a8e767c824d9535294e2f64b16c5edd97a268901b' >>>\

  **Linux my.remote.machine 4.18.0-372.26.1.0.1.el8_6.x86_64 #1 SMP Tue Sep 13 21:44:27 PDT 2022 x86_64 x86_64 x86_64 GNU/Linux**

**Exiting tunnel now...**
2022-09-19 12:09:12,761| INF | MainThrea/1374@sshtunnel | Closing all open connections...\
2022-09-19 12:09:12,761| DEB | MainThrea/1378@sshtunnel | Listening tunnels: 0.0.0.0:5000\
2022-09-19 12:09:12,761| INF | MainThrea/1453@sshtunnel | Closing ssh transport\
2022-09-19 12:09:12,761| INF | MainThrea/1458@sshtunnel | Shutting down tunnel: 0.0.0.0:5000 <> A.B.C.E:22 (up)\
2022-09-19 12:09:12,761| TRA | Thread-11/0327@sshtunnel | <<< IN #2 <-- ('127.0.0.1', 60446) recv is not ready <<<\
2022-09-19 12:09:12,761| TRA | Thread-11/0375@sshtunnel | #2 <-- ('127.0.0.1', 60446) connection closed.\
2022-09-19 12:09:12,776| INF |  Srv-5000/1439@sshtunnel | Tunnel: 0.0.0.0:5000 <> A.B.C.E:22 released\
2022-09-19 12:09:12,776| DEB | MainThrea/1477@sshtunnel | Transport is closed\
DONE...
```
The lines with KT = are debug print statements I put in paramiko's agent.py to print the type of key being tried.

I sent to the remote machine the UNIX command "uname -a" and returned the result back thru the tunnel (see the line my.remote.machine above). 

The code I used to establish the tunnel thru a gateway is:

```
with sshtunnel.open_tunnel(
        ssh_address_or_host=(GATEWAY_SERVER_IP, GATEWAY_PORT),
        ssh_username=SSH_UNAME,
        ssh_pkey=SSH_PKEY_FILENAME,
        ssh_private_key_password=SSH_PW,
        ssh_password=PW,
        local_bind_address=('0.0.0.0', 5000),
        remote_bind_address=(REMOTE_SERVER_IP, REMOTE_PORT),
        debug_level=DEBUG_LEVEL
        ) as tunnel:
    tunnel.check_tunnels()
    client = paramiko.SSHClient()
    client.load_system_host_keys()
    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    client.connect('127.0.0.1', 5000)
    stdin_, stdout_, stderr_ = client.exec_command('uname -a')
    time.sleep(5)
    stdout_.channel.recv_exit_status()
    lines = stdout_.readlines()
    for line in lines:
        print(line)
    print('Exiting tunnel now...')
    client.close()
```

I'll close this issue as it's not really a bug.

--Ed

**Comment by bskinn** [comment]
Glad you were able to figure it out -- thanks for updating here!
